File: algorithms/strings/run-length-encoding/typescript/runLengthEncoding.ts
export function runLengthEncoding(arr: number[]): number[] {
    if (arr.length === 0) return [];
    const result: number[] = [];
    let count = 1;
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] === arr[i - 1]) { count++; }
        else { result.push(arr[i - 1], count); count = 1; }
    }
    result.push(arr[arr.length - 1], count);
    return result;
}

File: algorithms/strings/run-length-encoding/java/RunLengthEncoding.java
import java.util.*;

public class RunLengthEncoding {
    public static int[] runLengthEncoding(int[] arr) {
        if (arr.length == 0) return new int[0];
        List<Integer> result = new ArrayList<>();
        int count = 1;
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] == arr[i - 1]) { count++; }
            else { result.add(arr[i - 1]); result.add(count); count = 1; }
        }
        result.add(arr[arr.length - 1]); result.add(count);
        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}
